src/eei.cpp

```diff
@@ -119,12 +118,96 @@ Literal EthereumInterface::callImport(Import *import, LiteralList& arguments) {
       memory.set<uint8_t>(j, src[i]);
     }
   }
+  
+  /*
+   * Memory Operations
+   */
 
-  void EthereumInterface::copyAddressToMemory(struct evm_address const& address, uint32_t dstoffset)
+  void EthereumInterface::loadMemory(const uint32_t srcOffset, uint8_t *dst, size_t length)
   {
-    heraAssert(memory.size() < (dstoffset + 20), "Out of bounds (destination) memory copy.");
-    for (uint32_t i = 0, j = dstoffset; j < (dstoffset + 20); i++, j++) {
-      memory.set<uint8_t>(j, address.bytes[i]);
-    }
+      heraAssert(length > 0, "Length must be nonzero");
+      heraAssert((srcOffset + length) > srcOffset, "Out of bounds (source) memory copy.");
+
+      for (uint32_t i = 0; i < length; ++i) {
+          dst[length - (i + 1)] = memory.get<uint8_t>(srcOffset + i);
+      }
+  }
+
+  void EthereumInterface::storeMemory(const uint32_t dstOffset, const uint8_t *src, size_t length)
+  {
+      heraAssert(length > 0, "Length must be nonzero");
+      heraAssert((dstOffset + length) > dstOffset, "Out of bounds (destination) memory copy.");
+      heraAssert(memory.size() < (dstOffset + length), "Out of bounds (destination) memory copy.");
+
+      for (uint32_t i = 0; i < length; ++i) {
+          memory.set<uint8_t>(dstOffset + length - (i + 1), *(src + i));
+      }
+  }
+
+  /*
+   * Memory Op Wrapper Functions
+   */
+
+  void EthereumInterface::loadUint256(const uint32_t srcOffset, evm_uint256be &dst)
+  {
+      loadMemory(srcOffset, dst.bytes, 32);
+  }
+
+  void EthereumInterface::storeUint256(const uint32_t dstOffset, const evm_uint256be &src)
+  {
+      storeMemory(dstOffset, src.bytes, 32);
+  }
+
+  void EthereumInterface::loadUint160(const uint32_t srcOffset, evm_address &dst)
+  {
+      loadMemory(srcOffset, dst.bytes, 20);
+  }
+
+  void EthereumInterface::storeUint160(const uint32_t dstOffset, const evm_address &src)
+  {
+      storeMemory(dstOffset, src.bytes, 20);
+  }
+
+  void EthereumInterface::loadUint128(const uint32_t srcOffset, evm_uint256be &dst)
+  {
+      loadMemory(srcOffset, dst.bytes, 16);
+  }
+
+  void EthereumInterface::storeUint128(const uint32_t dstOffset, const evm_uint256be &src)
+  {
+      heraAssert(!exceedsUint128(src), "Value at src cannot exceed 2^128-1");
+      storeMemory(dstOffset, src.bytes, 16);
+  }
+
+  /*
+   * Utilities
+   */
+
+  /* Checks if host supplied 256 bit value exceeds UINT128_MAX */
+  unsigned int EthereumInterface::exceedsUint128(const evm_uint256be &value)
```

Comment:
Will make this `static` (or at least `const`) as it is not reading from the class instance.